src/frequenz/client/reporting/_client.py

```diff
@@ -0,0 +1,289 @@
+            * microgrid_id: The microgrid ID.
+            * component_id: The component ID.
+            * metric: The metric name.
+            * value: The metric value.
+        """
+        data = self._data_pb
+        for mdata in data.microgrids:
+            mid = mdata.microgrid_id
+            for cdata in mdata.components:
+                cid = cdata.component_id
+                for msample in cdata.metric_samples:
+                    ts = msample.sampled_at.ToDatetime()
+                    met = Metric.from_proto(msample.metric).name
+                    value = (
+                        msample.sample.simple_metric.value
+                        if msample.sample.simple_metric
+                        else None
+                    )
+                    yield MetricSample(
+                        timestamp=ts,
+                        microgrid_id=mid,
+                        component_id=cid,
+                        metric=met,
+                        value=value,
+                    )
+
+    @property
+    def next_page_token(self) -> Any:
+        """Get the token for the next page of data.
+
+        Returns:
+            The token for the next page of data.
+        """
+        return self._data_pb.pagination_info.next_page_token
+
+
+class ReportingClient:
+    """A client for the Reporting service."""
+
+    def __init__(self, service_address: str):
+        """Create a new Reporting client.
+
+        Args:
+            service_address: The address of the Reporting service.
+        """
+        self._grpc_channel = grpcaio.insecure_channel(service_address)
+        self._stub = ReportingStub(self._grpc_channel)
+
+    # pylint: disable=too-many-arguments
+    async def iterate_single_metric(
+        self,
+        *,
+        microgrid_id: int,
+        component_id: int,
+        metric: Metric,
+        start_dt: datetime,
+        end_dt: datetime,
+        page_size: int = 1000,
+    ) -> AsyncIterator[Sample]:
+        """Iterate over the data for a single metric.
+
+        Args:
+            microgrid_id: The microgrid ID.
+            component_id: The component ID.
+            metric: The metric name.
+            start_dt: The start date and time.
+            end_dt: The end date and time.
+            page_size: The page size.
+
+        Yields:
+            A named tuple with the following fields:
+            * timestamp: The timestamp of the metric sample.
+            * value: The metric value.
+        """
+        async for page in self._iterate_components_data_pages(
+            microgrid_components=[(microgrid_id, [component_id])],
+            metrics=[metric],
+            start_dt=start_dt,
+            end_dt=end_dt,
+            page_size=page_size,
+        ):
+            for entry in page.iterate_metric_samples():
+                yield Sample(timestamp=entry.timestamp, value=entry.value)
+
+    # pylint: disable=too-many-arguments
+    async def _iterate_components_data_pages(
```

Comment:
I would probably expose this publicly, in case someone wants to get pages for some reason, I don't see a reason to hide it.

Even more, now with the code in front of my eyes, I think the best approach would be to `list_microgrid_components_data()` return an object with a few iterable properties:

```py
async for page in client.list_microgrid_components_data(...).pages:
    ...

async for sample in client.list_microgrid_components_data(...).samples:
    ...
```

So `list_microgrid_components_data()` could return a `Response` object with properties like (pseudo-code using the current method names):

```py
@property
async def samples(self) -> AsyncIterator[Sample]:
    return self.client._iterate_single_metric(...)

@property
async def pages(self) -> AsyncIterator[ComponentsDataPage]:
    return self.client._iterate_components_data_pages(...)
```

This design might need a bit more thought, so maybe we should wait on it and discuss it in a separate issue.

Reply:
* https://github.com/frequenz-floss/frequenz-client-base-python/discussions/41

Reply:
> I would probably expose this publicly, in case someone wants to get pages for some reason, I don't see a reason to hide it.

Honestly I don't see a reason to expose it now. If we realize this becomes important, it's easy to expose. Because I don't see a use-case where you would want the pages that you cannot achieve with the iterator. 

> Even more, now with the code in front of my eyes, I think the best approach would be to list_microgrid_components_data()

That could be an option. But do you have a good example where having pages are better than an iterator over the samples?

Reply:
> Because I don't see a use-case where you would want the pages that you cannot achieve with the iterator.

Performance. We tested this and iterating over items was considerably slower than going through pages.

Reply:
Again, maybe because of data is used when doing data science, at the end is the same for that use case, but I think the API client should be pretty flexible and general purpose. And again, I'm thinking more long term, not just about what we need right now. For the right now I agree it is fine and we can improve later, but if we agree about this, I would create issues about it so:
1. We don't forget
2. It is clear for everyone what's the future direction we want to take, so issues with that future direction can also be raised early if anyone sees any.

Reply:
https://github.com/frequenz-floss/frequenz-client-reporting-python/issues/31